packages/serverpod/lib/src/database/database_connection.dart

```diff
@@ -189,6 +220,81 @@ Current type was $T''');
     return list.cast<T>();
   }
 
+  Map<String, dynamic> _resolveRowDataHierarchy(
+    Include include,
+    Table table,
+    Map<String, Map<String, dynamic>> rawRow,
+    Map<String, dynamic> parentTableRow,
+    String prefix,
+  ) {
+    include.includes.forEach((relationField, relationInclude) {
+      if (relationInclude == null) return;
+
+      var rawRelationTableRow = rawRow[relationInclude.table.tableName];
+      if (rawRelationTableRow == null) return;
+
+      var includePrefix = _createIncludePrefix(prefix, relationField);
+
+      var nonPrefixedRelationTableRow = <String, dynamic>{};
+      for (var column in relationInclude.table.columns) {
+        var columnName = column.columnName;
+        var includePrefixedColumnName = '$includePrefix.${column.columnName}';
+
+        var columnData = rawRelationTableRow[includePrefixedColumnName];
+        if (columnData != null) {
+          nonPrefixedRelationTableRow[columnName] =
+              rawRelationTableRow[includePrefixedColumnName];
+        }
+      }
+
+      if (nonPrefixedRelationTableRow.isNotEmpty) {
+        parentTableRow[relationField] = _resolveRowDataHierarchy(
+          relationInclude,
+          relationInclude.table,
+          rawRow,
+          nonPrefixedRelationTableRow,
+          includePrefix,
+        );
+      }
+    });
+
+    return parentTableRow;
+  }
```

Comment:
Suggestion: Break parts of this into named private methods to make it easier to understand the steps.
Perhaps we can find a better name for the method it self?

Two thoughts:
I think it is potentially a little hard to understand what we are doing here, I'm not sure what the name really means, or why we need a recursive call.

I also think it can be misleading returning a value when we are also mutating the returned value from the input params.